A38-admin-interface-api.md

```diff
@@ -0,0 +1,200 @@
+
+1. The default unsecured pattern has caused endpoint security concern in Envoy (see [envoy#2763])(https://github.com/envoyproxy/envoy/issues/2763));
+2. gRPC as a library should not open additional listening port without explicitly application code.
+
+
+### Alternative Admin Interface API Designs
+
+Here are the usage snippets for other API options for Java:
+
+Option 3: Server class method
+```java
+server = ServerBuilder.forPort(50051)
+        .useTransportSecurity(certChainFile, privateKeyFile)
+        .addAdminService()
+        .build()
+        .start();
+server.awaitTermination();
+```
+
+Option 4: Create admin server
+```java
+server = AdminServerBuilder.forPort(50051)
+        .useTransportSecurity(certChainFile, privateKeyFile)
+        .build()
+        .start();
+server.awaitTermination();
+```
+
+Option 5: Create and start serving
+```java
+server = admin.StartAdminServer(50051, certChainFile, privateKeyFile)
+server.awaitTermination();
+```
+
+For option 4/5, they well achieved the ultimate goal of reducing the line-of-code changes, but they also terrible at satisfying any need for extension. For option 5, advanced users might want to play with the admin service like adding interceptors or change credential types, but even tiny amount of change means fallback to option 1 (existing API). For option 4, the `AdminServerBuilder` allows some flexibility, but it conflicts with `XdsServerBuilder` if users want to enable xDS, and it would be repetitive if we offers `AdminXdsServerBuilder` on the side.
+
+The drawback for option 3 is dependency management. To date, the gRPC official admin services Channelz/Reflection/Health are distributed as separated package to keep the main package focus as a network library. If we apply option 3, then we will void the separate effort.
+
+
+## Future Work
+
+### Console-to-Console Credentials
+
+To date, if we want to securely guard an endpoint, the recommended way is using a TLS cert or via xDS’s mTLS (if rolled out). With proper setup, they work fine for long-lasting debugging ports. However, both of them are overly complicated for ad-hoc debugging, and may unfortunately nudge people to use insecure connections. From users’ perspective, they just want to inspect an gRPC application in one console with the tool in another console.
+
+The problem has been solved by many other softwares, like Grafana uses username and password, iPython notebook prints a generated auth token to the console. The admin server serves sensitive traffic statistics that should be protected. With easy-to-use credentials, we are promoting protected connections, and it will prevent gRPC from following the Envoy’s admin interface design.
```

Comment:
So which "easy-to-use" credentials are being recommended here?

Reply:
Thanks for paying attention to this part. From my view, both a simple-auth-style check and an auth token works. The challenging part will be pushing the design through, and get it accepted by the security people... What do you think? Do you have a preference about how to secure an admin endpoint?

Reply:
Hmmm, difficult to answer. I agree we should have an easy-to-use mechanism to prevent fallback to plaintext. I like the Jupyter notebook approach where they print out the token once you run a command. So we need another dummy command/executable to get this token to use in the actual admin command?

Reply:
That would be out of the scope of this gRFC.

If we decided to add a new member to existing credential types, we will need to be committed to implement it across languages. On the other hand, this type of "credential" much like other OAuth tokens can be implemented via interceptors, so there might be a third way to add it as a gRPC add-on library under "grpc-ecosystem" organization.